runtime/src/snapshot_utils.rs

```diff
@@ -179,19 +220,104 @@ pub fn package_snapshot<P: AsRef<Path>, Q: AsRef<Path>>(
     archive_format: ArchiveFormat,
     snapshot_version: SnapshotVersion,
     hash_for_testing: Option<Hash>,
-) -> Result<AccountsPackagePre> {
-    // Hard link all the snapshots we need for this package
+) -> Result<AccountsPackagePre>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    info!(
+        "Package snapshot for bank: {} has {} account storage entries",
+        bank.slot(),
+        snapshot_storages.len()
+    );
+
     let snapshot_tmpdir = tempfile::Builder::new()
         .prefix(&format!("{}{}-", TMP_SNAPSHOT_PREFIX, bank.slot()))
         .tempdir_in(snapshot_path)?;
 
-    // Create a snapshot package
+    do_package_snapshot(
+        bank,
+        snapshot_files,
+        status_cache_slot_deltas,
+        snapshot_package_output_path,
+        snapshot_storages,
+        archive_format,
+        snapshot_version,
+        hash_for_testing,
+        snapshot_tmpdir,
+    )
+}
+
+/// Package up slot snapshot files, snapshot storages, and slot deltas for an incremental snapshot.
+#[allow(clippy::too_many_arguments)]
+fn package_incremental_snapshot<P, Q>(
+    bank: &Bank,
+    incremental_snapshot_base_slot: Slot,
+    snapshot_files: &SlotSnapshotPaths,
+    snapshot_path: Q,
+    status_cache_slot_deltas: Vec<BankSlotDelta>,
+    snapshot_package_output_path: P,
+    snapshot_storages: SnapshotStorages,
+    archive_format: ArchiveFormat,
+    snapshot_version: SnapshotVersion,
+    hash_for_testing: Option<Hash>,
+) -> Result<AccountsPackagePre>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
     info!(
-        "Snapshot for bank: {} has {} account storage entries",
+        "Package incremental snapshot for bank {} (from base slot {}) has {} account storage entries",
         bank.slot(),
+        incremental_snapshot_base_slot,
         snapshot_storages.len()
     );
 
+    assert!(
+        snapshot_storages.iter().all(|storage| storage
+            .iter()
+            .all(|entry| entry.slot() > incremental_snapshot_base_slot)),
+        "Incremental snapshot package must only contain storage entries where slot > incremental snapshot base slot (i.e. full snapshot slot)!"
+    );
+
+    let snapshot_tmpdir = tempfile::Builder::new()
+        .prefix(&format!(
+            "{}{}-{}-",
+            TMP_INCREMENTAL_SNAPSHOT_PREFIX,
+            incremental_snapshot_base_slot,
+            bank.slot()
+        ))
+        .tempdir_in(snapshot_path)?;
+
+    do_package_snapshot(
+        bank,
+        snapshot_files,
+        status_cache_slot_deltas,
+        snapshot_package_output_path,
+        snapshot_storages,
+        archive_format,
+        snapshot_version,
+        hash_for_testing,
```

Comment:
Since all of these arguments are shared between `do_package_snapshot`, `package_incremental_snapshot`, and `package_snapshot`, it might be cleaner to wrap these up in an argument struct

Reply:
While implementing this, I ended up feeling like I didn't actually help anything, just move around where I was writing all the parameters, and even adding a bit of work to create the struct and then destructure it later. How strongly would you like this change to go in?

Reply:
Heh, looking at it again, I think `package_incremental_snapshot` and  `package_snapshot` may just be better off as constructors for `AccountsPackagePre`, where logic like https://github.com/solana-labs/solana/pull/18504/files#diff-c957621ae265a12e03f7b9c5656e7aa74137223e603cc3cfcc2828dda1af5479R398-R405 could be moved into a constructor.

Also has the added benefit of offloading from the `snapshot_utils` module as it's growing rather large.

Reply:
After talking with @carllin, these changes will not be part of this PR. I'll likely pull it into another change later. Basically just pulling the TempDir into constructors for AccountsPackagePre.

Reply:
New issue created to track this work: #18815